cuML/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,93 @@
+# cuML developer guide
+This document summarizes rules and best practices for contributions to C++ component cuML of RAPIDS/cuml.
+
+# General
+
+# Thread safety
+
+# Coding style
+
+# Error handling
+All calls to CUDA APIs should be done via the provided helper macros `CUDA_CHECK`, `CUBLAS_CHECK` and `CUSOLVER_CHECK`. Those macros take care of checking the return values of the used API calls and generate an exception in case the command was no successful. In case an exception needs to be avoided, e.g. when implementing a destructor, `CUDA_CHECK_NO_THROW`, `CUBLAS_CHECK_NO_THROW ` and `CUSOLVER_CHECK_NO_THROW ` (currently not available, see https://github.com/rapidsai/cuml/issues/229) should be used. Those macros will only log the error but do not throw an exception.
+
+# Logging
+Add once https://github.com/rapidsai/cuml/issues/100 is addressed.
+
+# Device and Host memory allocations
+To enable `libcuml.so` users to control how memory for temporary data is allocated device memory should only be allocated via the allocator provided:
+```cpp
+template<typename T>
+void foo(ML::cumlHandle* handle, cudaStream_t stream, ... )
+{
+    T* temp_h = handle->getDeviceAllocator()->allocate(n*sizeof(T), stream);
```

Comment:
@jrhemstad IIUC, page locked memory is what is being proposed with the host_allocator concept below.

Reply:
Liked your suggestion about reducing verbosity.

Reply:
Interesting idea regarding the specialized allocators for Unified Memory and pinned host memory. However before we add these kind of things I would prefer to see the use case first. Currently I know that we need temporary host memory and temporary device memory. The interface is sufficient to support that and already gives the user of cuML quite some flexibility, e.g. he could pass in an allocator that allocates pinned host memory or one that allocates pageable host memory (current default is pinned host memory). Similar for temporary device memory he could use Unified Memory, pinned host memory or pinned device memory.
Regarding the less verbose API: I actually had that in an earlier version I tried, but preferred a cleaner interface that does not duplicate functionality between classes. Also the `ML::device_buffer` and `ML::host_buffer` should be preferred for allocations IMHO. A verbose interface make the alternative less attractive. However I do not have a very strong opinion on this one so I am open to provide a device_allocate and host_allocate in cumlHandle.

Reply:
> I would prefer to see the use case first.

I have no idea what the use-case may be :) This is my first time looking at cuML code. That said, we can never predict the future, so it's safest to assume that if something _can_ happen (like wanting to use UVM), it probably _will_ at some point and the design should be able to handle that.

> Similar for temporary device memory he could use Unified Memory, pinned host memory or pinned device memory.

This would require the user defining their own custom allocator then? I think that's fine, just making sure I understand.

> Also the `ML::device_buffer` and `ML::host_buffer` should be preferred for allocations IMHO. A verbose interface make the alternative less attractive.

100% agreed that RAII constructs should be preferred. I had a sneaking suspicion that you left the `malloc/free`-like API verbose to discourage people from using it :)

Reply:
> > I would prefer to see the use case first.

> I have no idea what the use-case may be :) This is my first time looking at cuML code. That said,  we can never predict the future, so it's safest to assume that if something _can_ happen (like wanting to use UVM), it probably _will_ at some point and the design should be able to handle  that.

I would say the current design is flexible enough for that. Once the need comes up we can add allocation routines to the proposed allocator interfaces with default implementations, e.g. for manged memory calling through to `cudaMallocManaged()`. Since we won't need to make allocations with a new type of memory before that happens it won't require to touch old cuML code. And cuML users affected be the default implementation can easily override the default one. Does that seem reasonable to you?

> > Similar for temporary device memory he could use Unified Memory, pinned host memory or pinned device memory.

> This would require the user defining their own custom allocator then? I think that's fine, just making sure I understand.

Yes. A example for a custome allocator that falls back to is 
```cpp
class cachingDeviceAllocator : public ML::deviceAllocator {
public:
    cachingDeviceAllocator()
          , _allocator(8, 3, cub::CachingDeviceAllocator::INVALID_BIN, cub::CachingDeviceAllocator::INVALID_SIZE)
    {}

    virtual void* allocate( std::size_t n, cudaStream_t stream ) {
        void* ptr = 0;
        _allocator.DeviceAllocate( &ptr, n, stream );
        return ptr;
    }

    virtual void deallocate( void* p, std::size_t, cudaStream_t ) {
        _allocator.DeviceFree(p);
    }
    
private:
    cub::CachingDeviceAllocator _allocator;
};
```
That would be passed to cuML like this
```cpp
std::shared_ptr<ML::rmmAllocatorAdapter> allocator( new ML::rmmAllocatorAdapter() );
cumlHandle->setDeviceAllocator( allocator );
```

Reply:
Let's just keep the current verbose API as it is. Depending on which one of these interfaces (device_buffer or allocate calls) folks end up preferring, we can then revisit the API restructuring.